modules/tinymce/src/themes/silver/main/ts/ui/context/ContextToolbarLookup.ts

```diff
@@ -5,19 +5,63 @@
 import { Node as DomNode } from '@ephox/dom-globals';
 import { Arr, Option } from '@ephox/katamari';
 import { Compare, Element, TransformFind } from '@ephox/sugar';
 import Editor from 'tinymce/core/api/Editor';
-
+import { ContextTypes } from '../../ContextToolbar';
 import { ScopedToolbars } from './ContextToolbarScopes';
 
-export type LookupResult = { toolbarApi: Toolbar.ContextToolbar | Toolbar.ContextForm, elem: Element };
+export type LookupResult = { toolbars: ContextTypes[], elem: Element };
+type MatchResult = { contextToolbars: ContextTypes[], contextForms: ContextTypes[] };
+
+const matchTargetWith = (elem: Element, candidates: ContextTypes[]): MatchResult => {
+  const ctxs = Arr.filter(candidates, (toolbarApi) => toolbarApi.predicate(elem.dom()));
+  const { pass, fail } = Arr.partition(ctxs, (t) => t.type === 'contexttoolbar');
+  return { contextToolbars: pass, contextForms: fail };
+};
+
+const filterToolbarsByPosition = (toolbars: ContextTypes[]): ContextTypes[] => {
+  if (toolbars.length <= 1) {
+    return toolbars;
+  } else {
+    const findPosition = (value) => Arr.find(toolbars, (t) => t.position === value);
 
-const matchTargetWith = (elem: Element, toolbars: Array<Toolbar.ContextToolbar | Toolbar.ContextForm>): Option<LookupResult> => {
-  return Arr.findMap(toolbars, (toolbarApi) =>
-    toolbarApi.predicate(elem.dom()) ? Option.some({ toolbarApi, elem }) : Option.none());
+    // prioritise position by 'selection' -> 'node' -> 'line'
+    const basePosition = findPosition('selection')
+        .orThunk(() => findPosition('node'))
+        .orThunk(() => findPosition('line'))
+        .map((t) => t.position);
+    return basePosition.fold(
+      () => [],
+      (pos) => Arr.filter(toolbars, (t) => t.position === pos)
+    );
+  }
+};
+
+const matchStartNode = (elem: Element, nodeCandidates: ContextTypes[], editorCandidates: ContextTypes[]): Option<LookupResult> => {
+  // requirements:
+  // 1. prioritise context forms over context menus
+  // 2. prioritise node scoped over editor scoped context forms
+  // 3. only show max 1 context form
+  // 4. concatenate all available context toolbars if no context form
+
+  const nodeMatches = matchTargetWith(elem, nodeCandidates);
+
+  if (nodeMatches.contextForms.length > 0) {
+    return Option.some({ elem, toolbars: [nodeMatches.contextForms[0]] });
+  } else {
+    const editorMatches = matchTargetWith(elem, editorCandidates);
+
+    if (editorMatches.contextForms.length > 0) {
+      return Option.some({ elem, toolbars: [editorMatches.contextForms[0]] });
+    } else if (nodeMatches.contextToolbars.length > 0 || editorMatches.contextToolbars.length > 0) {
+      const toolbars = filterToolbarsByPosition(nodeMatches.contextToolbars.concat(editorMatches.contextToolbars));
+      return Option.some({ elem, toolbars });
+    } else {
+      return Option.none();
+    }
+  }
 };
 
 const lookup = (scopes: ScopedToolbars, editor: Editor): Option<LookupResult> => {
```

Comment:
@lnewson I finally merged in your changes (e.g. is OutsideRoot). Been a while though so not entirely sure I sorted out all the isRoot checks and such correctly though. Please check this?

Reply:
I added a test to cover the cases, so if it's passing you should be fine 👍 I'll take a proper look tomorrow though just to be sure.

Reply:
Looks good to me, you will have to update one of the assertions in the new tests though since we're now merging the toolbars.

```
--- modules/tinymce/src/themes/silver/test/ts/browser/editor/contexttoolbar/ContextToolbarLookupTest.ts	(revision 9e6393b75a1280bde5d6b42f35153eb4835e45c2)
+++ modules/tinymce/src/themes/silver/test/ts/browser/editor/contexttoolbar/ContextToolbarLookupTest.ts	(date 1584583916643)
@@ -50,7 +50,7 @@
           sResetNames(),
           tinyApis.sSetCursor([0, 0, 0], 1),
           UiFinder.sWaitForVisible('Waiting for node toolbar to appear', Body.body(), '.tox-tbtn:contains(Node)'),
-          sAssertNames(['a'], [], [])
+          sAssertNames(['a'], ['a'], ['a'])
         ]),
 
         Log.stepsAsStep('TINY-4571', 'Context toolbar parent node scope lookup', [
```

Reply:
You might also want to add some other tests to that for your new logic 🤷‍♀

Reply:
Yeah, need to do that yet...